consensus/service/src/byzantine_ledger/worker.rs

```diff
@@ -418,183 +404,209 @@ impl<
+            // Broadcast resolved messages.
+            for (consensus_msg, from_responder_id) in &resolved {
                 self.broadcaster
                     .lock()
                     .expect("mutex poisoned")
                     .broadcast_consensus_msg(consensus_msg.as_ref(), &from_responder_id);
+            }
 
-                // Unclear if this helps with anything, so it is disabled for now.
-                /*
-                // See if this message has any values to incorporate into our pending
-                // values. This is safe since we only grab values we think are
-                // potentially valid.
-                let mut grabbed = 0;
-                if let Some(voted_or_accepted_nominated) = scp_msg.votes_or_accepts_nominated() {
-                    for value in voted_or_accepted_nominated {
-                        if !self.pending_values_map.contains(&value) && self.tx_cache.validate_tx_by_hash(&value).is_ok() {
-                            self.pending_values.insert(value.clone());
-                            self.pending_values_map.insert(value, Instant::now()? not sure if this is reasonable);
-                            grabbed += 1;
-                        }
-                    }
-                }
-
-                if grabbed > 0 {
-                    log::debug!(self.logger, "Grabbed {} extra pending values from SCP traffic", grabbed);
-                }
-                */
+            let scp_msgs: Vec<Msg<_>> = resolved
+                .into_iter()
+                .map(|(consensus_msg, _)| consensus_msg.scp_msg().clone())
+                .collect();
 
-                // Pass message to the scp layer.
-                match self.scp.handle(scp_msg) {
-                    Ok(msg_opt) => {
-                        if let Some(msg) = msg_opt {
-                            (self.send_scp_message)(msg);
-                        }
-                    }
-                    Err(err) => {
-                        log::error!(
-                            self.logger,
-                            "Failed handling message {:?}: {:?}",
-                            scp_msg,
-                            err
-                        );
+            match self.scp.handle_messages(scp_msgs) {
+                Ok(outgoing_msgs) => {
+                    for msg in outgoing_msgs {
+                        (self.send_scp_message)(msg);
                     }
                 }
+                Err(err) => {
+                    log::error!(self.logger, "Failed handling messages: {:?}", err);
+                }
             }
         }
     }
 
     fn attempt_complete_cur_slot(&mut self) {
         // See if we have externalized values for the current slot.
-        let ext_vals = self.scp.get_externalized_values(self.cur_slot);
-        if ext_vals.is_empty() {
-            return;
-        }
+        match self.scp.get_externalized_values(self.current_slot_index) {
```

Comment:
`if let Some(ext_vals) = self.scp.get_externalized_values(self.current_slot_index) {` is more elegant, imo (and is nested one level less)